llvm/tools/llvm-snippy/README.md

```diff
@@ -0,0 +1,252 @@
+> cmake --build release/build
+> cmake --install release/build
+```
+
+We haven't open-sourced LIT tests yet. When we do that, there will also be instructions here on how to test the generator.
+
+These steps should get llvm-snippy up and running on your system. If you encounter any issues or have questions, feel free to reach out.
+
+# Quick start guide
+
+In this section, we will create a configuration file, run the generator, and examine the generated results.
+
+## Creating the Configuration File
+
+In this section, we'll guide you through the process of creating a configuration file for llvm-snippy. 
+
+Create a file named layout.yaml with the following content:
+
+```
+options:
+  march: "riscv64-linux-gnu"
+  model-plugin: "None"
+  num-instrs: 10000
+  o: "snippet.elf"
+  init-regs-in-elf: true
+  honor-target-abi: true
+  stack-size: 1024
+  last-instr: "RET"
+
+sections:
+  - name:      text
+    VMA:       0x210000
+    SIZE:      0x100000
+    LMA:       0x210000
+    ACCESS:    rx
+  - name:      data
+    VMA:       0x100000
+    SIZE:      0x100000
+    LMA:       0x100000
+    ACCESS:    rw
+
+histogram:
+    - [ADD, 1.0]
+    - [ADDI, 1.0]
+    - [SUB, 1.0]
+    - [SRA, 1.0]
+    - [SRAI, 1.0]
+    - [SRL, 1.0]
+    - [SRLI, 1.0]
+    - [SLL, 1.0]
+    - [SLLI, 1.0]
+    - [AND, 1.0]
+    - [ANDI, 1.0]
+    - [OR, 1.0]
+    - [ORI, 1.0]
+    - [XOR, 1.0]
+    - [XORI, 1.0]
+    - [LW, 10.0]
+    - [SW, 10.0]
+
+access-ranges:
+   - start: 0x100000
+     size: 0x10000
+     stride: 16
+     first-offset: 0
+     last-offset: 2
+```
+
+Let's briefly explain the content of the layout.yaml file:
+
+### options
+
+This section allows you to set various command-line options. They can also be set directly in the command line. Here's a breakdown of some options:
+
+- march: Target architecture (e.g., "riscv64-linux-gnu").
+- model-plugin: Model plugin to use (e.g., "None").
+- num-instrs: Number of instructions.
+- o: Base name for the output file.
+- init-regs-in-elf: Initialize non-fixed registers in the ELF.
+
+```
+march: "riscv64-linux-gnu"
+```
+
+The llvm-snippy generator potentially supports a variety of architectures. Currently, the only working backend is RISCV.
+
+```
+model-plugin: "None"
+```
+
+The llvm-snippy generator also supports various models through a unified interface. We plan to open-source SAIL-based and spike-based models that are already operational within Syntacore. Currently, the primary mode is without a model unless you write and integrate your own.
+
+```
+num-instrs: 10000
+o: "snippet.elf"
+```
+
+These options specify the number of instructions and the base name of the output file. You can set num-instrs to "all" and the entire section will be filled.
+
+```
+init-regs-in-elf: true
+```
+
+It initializes all unfixed registers with random values at the beginning of the snippet. This option helps avoid non-determinism due to different register settings before a function call.
```

Comment:
unfixed? Are you sure that it's the best term? Personally I don't understand it :). I suggest using available, reserved and so on.

Reply:
non-reserved